mino/minows/session.go

```diff
@@ -0,0 +1,250 @@
+package minows
+
+import (
+	"context"
+	"encoding/gob"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/serde"
+	"golang.org/x/xerrors"
+	"io"
+	"sync"
+)
+
+var ErrWrongAddressType = xerrors.New("wrong address type")
+var ErrNotPlayer = xerrors.New("not player")
+
+// MaxUnreadAllowed Maximum number of unread messages allowed
+// in orchestrator's incoming message buffer before pausing relaying
+const MaxUnreadAllowed = 1e3
+
+type Forward struct {
+	Packet
+	Destination []byte
+}
+
+type orchestrator struct {
+	logger zerolog.Logger
+
+	myAddr orchestratorAddr
+	rpc    rpc
+	// Connects to the participants
+	outs map[peer.ID]*gob.Encoder
+	in   chan Packet
+}
+
+func (o orchestrator) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+	return doSend(addrs, msg, o.send, o.logger)
+}
+
+func (o orchestrator) Recv(ctx context.Context) (mino.Address, serde.Message, error) {
+	unpack := unpacker(o.rpc.mino.GetAddressFactory(), o.rpc.factory, o.rpc.context)
+	return doReceive(ctx, o.in, unpack, o.logger)
+}
+
+func (o orchestrator) send(addr mino.Address, msg serde.Message) error {
+	var unwrapped address
+	switch a := addr.(type) {
+	case address:
+		unwrapped = a
+	case orchestratorAddr:
+		unwrapped = a.address
+	default:
+		return xerrors.Errorf("%v: %T", ErrWrongAddressType, addr)
+	}
+
+	encoder, ok := o.outs[unwrapped.identity]
+	if !ok {
+		return xerrors.Errorf("%v: %v", ErrNotPlayer, addr)
+	}
+	src, err := o.myAddr.MarshalText()
+	if err != nil {
+		return xerrors.Errorf("could not marshal address: %v", err)
+	}
+	payload, err := msg.Serialize(o.rpc.context)
+	if err != nil {
+		return xerrors.Errorf("could not serialize message: %v", err)
+	}
+
+	err = encoder.Encode(&Packet{Source: src, Payload: payload})
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
```

Comment:
this would be so much simpler with serde, and not mixed with the `session` code.